<a href="https://colab.research.google.com/github/byui-cse/cse380-notebooks/blob/master/10_3_About_Patterns_and_Probabilities.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# About Patterns and Probabilities
## Class Directed Learning
### Due: Tuesday, 9 March 2021, 11:59 pm

## TODO Explore and Wonder

Regarding spanning trees of ladder graphs:

What is the probability ($\lim_{n \rightarrow \infty} P(n)$) that a randomly-selected spanning tree of an $n$-rung ladder graph contains the bottom rung?

| n | P(n)  |
|---|------:|
| 1 |      1|
| 2 |    .75|
| 3 |  .7333|
| 4 |  .7321|
| 5 |  .7320|


$\lim_{n \rightarrow \infty} P(n) = 0.73205$

### Recurrence Relations

Let $f(n) =$ NSTIBR$(n)$:

$f(n) = 4f(n-1) - f(n-2)$ for $n > 1$;

$f(0) = 1$,

$f(1) = 1$.

Let $g(n) =$ NST$(n)$:

$g(n) = 4g(n-1) - g(n-2)$ for $n > 1$;

$g(0) = 0$,

$g(1) = 1$.

In [21]:
def nstibr(n):
    f = 0
    if n > 1:
        f += 4 * nstibr(n-1) - nstibr(n-2)
    if (n == 1 or n == 0):
        f = 1
    return f

def nst(n):
    g = 0
    if n > 1:
        g += 4 * nst(n-1) - nst(n-2)
    if (n == 1):
        g = 1
    elif (n == 0):
        g = 0
    return g

def pstibr(n):
    if (n > 0):
        return nstibr(n) / nst(n)
    else:
        return 0

In [22]:
print("n\tf(n)\tg(n)\tP(STIBR)")
for i in range(0, 14):
    print(str(i) + "\t" + str(nstibr(i)) + "\t" + str(nst(i)) + "\t" + str(pstibr(i)))

n	f(n)	g(n)	P(STIBR)
0	1	0	0
1	1	1	1.0
2	3	4	0.75
3	11	15	0.7333333333333333
4	41	56	0.7321428571428571
5	153	209	0.7320574162679426
6	571	780	0.732051282051282
7	2131	2911	0.7320508416351769
8	7953	10864	0.7320508100147275
9	29681	40545	0.7320508077444814
10	110771	151316	0.7320508075814851
11	413403	564719	0.7320508075697825
12	1542841	2107560	0.7320508075689423
13	5757961	7865521	0.732050807568882


#### TODO Check Recurrences

Check the calculations in this table (maybe write recursive functions) to verify they are accurate.

Do they match what you found yesterday in your DPC?

| n | f(n) | f(n-1) | 4f(n-1)  | f(n - 2) | diff |
|---|-----:|-------:|---------:|---------:|-----:|
| 0 |    1 |    N/A |      N/A |      N/A |  N/A |
| 1 |    1 |      1 |        4 |      N/A |  N/A |
| 2 |    3 |      1 |        4 |        1 |    3 |
| 3 |   11 |      3 |       12 |        1 |   11 |
| 4 |   41 |     11 |       44 |        3 |   41 |
| 5 |  153 |     41 |      164 |       11 |  153 |

| n | g(n) | g(n-1) | 4g(n-1)  | g(n - 2) | diff |
|---|-----:|-------:|---------:|---------:|-----:|
| 0 |    0 |    N/A |      N/A |      N/A |  N/A |
| 1 |    1 |      0 |        0 |      N/A |  N/A |
| 2 |    4 |      1 |        4 |        0 |    4 |
| 3 |   15 |      4 |       16 |        1 |   15 |
| 4 |   56 |     15 |       60 |        4 |   56 |
| 5 |  209 |     56 |      224 |       15 |  209 |

### TODO Find closed-form formulas

Can you find closed-form formulas for $f(n)$ and $g(n)$?

A closed-form formula expressing these functions in terms of operations on $n$, **without** referring to previous calculated values of the functions.

#### Hint:

http://www.ist.tugraz.at/aichholzer/teaching/eca/spanning_trees_in_ladders.pdf

#### Possibly Illuminating Calculations

Remember continued fractions?

In [23]:
from fractions import Fraction as frac

def contfrac2frac(seq):
    """Convert the simple continued fraction in `seq`
       into a fraction with numerator num and denominator den.
    """
    num, den = 1, 0
    for u in reversed(seq):
        num, den = den + num * u, num
    return frac(num, den)

def frac2contfrac(f):
    """Build the simple continued fraction expansion of fraction f.
    """
    seq = []
    frac2contfrac_rec(f, seq)
    return seq

def frac2contfrac_rec(f, seq):
    n = f.numerator
    d = f.denominator
    if d != 0:
        seq.append(n // d)
        if n % d != 0:
            frac2contfrac_rec(frac(d, n % d), seq)

def eval_frac(f):
    """Evaluate the fraction f as a float.
    """
    return f.numerator / f.denominator

In [24]:
from math import sqrt

value = sqrt(3) - 1
value_as_cf = frac2contfrac(frac.from_float(value))[:21]
cf_to_value = contfrac2frac(value_as_cf)
print(value, value_as_cf, cf_to_value)
print(eval_frac(cf_to_value))

0.7320508075688772 [0, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2] 302632/413403
0.732050807565499


In [25]:
from pprint import pprint

pprint(list(map(lambda n: contfrac2frac(value_as_cf[:n]), range(20, 5, -1))))

[Fraction(110771, 151316),
 Fraction(81090, 110771),
 Fraction(29681, 40545),
 Fraction(21728, 29681),
 Fraction(7953, 10864),
 Fraction(5822, 7953),
 Fraction(2131, 2911),
 Fraction(1560, 2131),
 Fraction(571, 780),
 Fraction(418, 571),
 Fraction(153, 209),
 Fraction(112, 153),
 Fraction(41, 56),
 Fraction(30, 41),
 Fraction(11, 15)]


g(n) = $\frac{1}{2\sqrt{3}}(2 + \sqrt{3})^n - \frac{1}{2\sqrt{3}}(2 - \sqrt{3})^n$

f(n) = $g(n) - g(n-1)$

In [26]:
def nst2(n):
    return int(1/(2*sqrt(3)) * (2 + sqrt(3)) ** n - 1/(2*sqrt(3)) * (2 - sqrt(3)) ** n)

def nstibr2(n):
    if (n == 0):
        return 0
    else:
        return nst2(n) - nst2(n-1)

0
1
3
11
41
153


In [35]:
# As a sanity check, let's get the first few calculations. 

print("n\tf(n)\tg(n)\tP(STIBR)")
print("0\t0\t1\t0")
for i in range(1, 14):
    print(str(i) + "\t" + str(nstibr2(i)) + "\t" + str(nst2(i)) + "\t" + str(eval_frac(frac(nstibr2(i), nst2(i)))))

# The results are consistent with what we got with the recursive functions, and we again see that the limit of P(STIBR) as n approaches inf is ~ 0.73205

n	f(n)	g(n)	P(STIBR)
0	0	1	0
1	1	1	1.0
2	3	4	0.75
3	11	15	0.7333333333333333
4	41	56	0.7321428571428571
5	153	209	0.7320574162679426
6	571	780	0.732051282051282
7	2131	2911	0.7320508416351769
8	7952	10863	0.7320261437908496
9	29682	40545	0.7320754716981132
10	110770	151315	0.7320490367775832
11	413403	564718	0.7320521038819375
12	1542841	2107559	0.7320511549142871
13	5757961	7865520	0.7320509006397543
